module: add crystal module

angelolu opened the pull request:
This commit adds a module called crystal and an optional invalid command
instance variable/handle to the Module class. The handle was added to
process messages where the second token is invalid or is the actual
actionable argument/option for simple modules.

The crystal module, requested by anoooooooooooooooooka, takes the input
number and sends back a message after doing some simple math on the
number. The application of this module is unknown; presumably it's for
some waste-of-time gacha game.

This commit is untested.

angelolu committed:
module: add crystal module

This commit adds a module called crystal and an optional invalid command
instance variable/handle to the Module class. The handle was added to
process messages where the second token is invalid or is the actual
actionable argument/option for simple modules.

The crystal module, requested by anoooooooooooooooooka, takes the input
number and sends back a message after doing some simple math on the
number. The application of this module is unknown; presumably it's for
some waste-of-time gacha game.

This commit is untested.

angelolu commented:
Not sure how to test/run this code to be honest. Please DM me.

ianw3214 commented on main.py:
```diff
@@ -22,6 +22,8 @@ def initModules():
     addModule(plant)
     coin = app.modules.coin.CoinModule(client)
     addModule(coin)
+    crystal = app.modules.crystal.CrystalModule(client)
```
You have to import the file at the top as well 🌾

angelolu replied:
Done. Fixed in b27df0c.

ianw3214 commented on app/modules/crystal.py:
```diff
@@ -0,0 +1,19 @@
+    def __init__(self, client):
+        app.module.Module.__init__(self, client, "crystal", self.calculate)
+
+    async def calculate(self, rawMessage, tokens):
```
Why use an invalid command handler to calculate crystals? Ideally an invalid command handler would just give the user some sort of feedback that their command doesn't work imo.

I think it would be better if this was handled by some sort of "default" handler, so that when no specific commands are passed it will run the default process. Whether the way to do that is to just rename `invalidCommandHandler` to `defaultCommandHandler` or making an entirely separate default handler is up to you.

angelolu replied:
Thank you for the feedback. I totally agree. Unfortunately, I blanked on the word "default" when I was writing this 🤷. I have renamed `invalidCommandHandler` to `defaultMessageHandler` in b27df0c. Please resolve this conversation if that is sufficient.

angelolu committed:
module: refactor invalidCommandHandler

This commit refactors the class instance variable invalidCommandHandler
to defaultMessageHandler to better communicate the intent of the
variable: an optional handler so modules can catch messages without a
registered command.

This commit also adds an import in main.py that was missed.

Tested locally on Python 3.8.10, Microsoft Windows [Version
10.0.22000.258].

RE: ianw3214/Beepbop/pull/1

IanWang2 reviewed (approved):
poggers

ianw3214 reviewed (approved):
poggers

angelolu commented:
Thank you for the approvals. Ready for merge at your convenience.
